### ISSUE TITLE
Revisit the question of validating operation in the query planner

### ISSUE THREAD

**Issue by pcmanus** [apollographql/federation#2258]
The query planner does not do a full validation of the query passed to it, but instead assumes that the query is valid. Note that plenty of invalid queries will result in the query planner complaining, because most invalid things gets directly in the way of the planning, but a few things are just ignored. For instance, the planner does not check for unused fragments. Or, maybe more importantly, the planner does not validate that a given selection does not query the same field with different arguments without aliases. Meaning that if one sends a query like:
```graphql
f(arg: 42) {
  a
}
f(arg: 24) {
  b
}
```
then that query is invalid for graphQL (you should alias one of the `f`, but the query planner will not reject it. Instead, it kind of assume that "response names" are unique, and will end up treating the query as:
```graphql
f(arg: 42) {
  a
  b
}
```
but the detail really shouldn't matter here, because the point is that the query is invalid in the first place and should not have been passed to the planner.

Note that this assumption dates back from federation 1, whose planner also behaves this way, and I believe the rational was to simply avoid wasting re-validating the query (for performance reason) when it's usually already validated by Apollo Server (which then only pass valid queries down to the gateway and planner).

However, it's not exactly impossible to bypass that validation. After all, the [gateway `executor`](https://github.com/apollographql/federation/blob/main/gateway-js/src/index.ts#L743) within which planning happens is public. Not to mention that external tooling may well use the planner directly, without even building a gateway. And we don't exactly document clearly that the queries passed to said gateway executor (or the planner directly) should have been validated beforehand. And as mentioned above, this could result in unexpected result pretty silently, and be quite hard to debug.

Long story short, we should probably consider making the planner a bit more defensive and add a few missing validation of the operation. At least the ones like the one described above that have genuine impact. It's unclear doing so would have all that much impact performance wise, but we can check. If the impact is noticeable, maybe we can add full validation, but allow to disable it optionally, and ensure that when the gateway is wired with Apollo Server and we're sure the query has been validated, then we skip the planner re-validation.
